[로또] 슬로스 미션 제출합니다.

soojinroh opened the pull request.

indiveloper committed:
feat 기능 구현 목록 작성

indiveloper committed:
feat LottoNumber 클래스 구현

1부터 45까지의 로또 번호 하나 하나를 의미하는 LottoNumber 클래스 작성. 1부터 45까지의 숫자만 허용하고 이외의 숫자가 입력될 경우 예외를 발생시킴.

indiveloper committed:
refactor EXCEPTION_MESSAGE 문구 수정

EXCEPTION_MESSAGE 문구 수정

indiveloper committed:
feat LottoNumbers 클래스 구현

1부터 45까지의 값을 가지고 있는 LottoNumber 객체 45개를 할당 받은 Map의 일급 컬렉션. static으로 구현하여 프로그램 내에서 재사용 할 수 있게 함.

indiveloper committed:
feat AutoLottoGenerator 구현

랜덤한 LottoNumber객체 6개로 이루어진 List<LottoNumber>를 생성해 반환하는 AutoLottoGenerator 구현

indiveloper committed:
feat ManualLottoGenerator 구현

AutoLottoGenerator는 랜덤하게 번호를 선택해 List<LottoNumber>를 생성한 반면, ManualLottoGenerator는 사용자로부터 입력 받은 6개의 중복되지 않는 int로 이루어진 List를 입력 받아 List<LottoNumber>를 생성해 리턴함

indiveloper committed:
refactor 변수명 변경

allLottoNumbers를 lottoNumbers로 변수명 변경

indiveloper committed:
feat Money 클래스 구현

로또 구입 금액을 의미하는 Money 클래스 구현.
1,000원 이상 100,000원 이하의 금액만 입력 가능하다.
1,000원을 기준으로 입력해야 한다. 100원, 10원, 1원 단위가 입력되었을 경우 예외를 던진다.

indiveloper committed:
refactor 변수명 변경

indiveloper committed:
refactor 클래스명 변경 및 equals, hashCode 오버라이드

indiveloper committed:
feat Lotto 클래스 구현

- 로또 한 게임(6개의 번호를 가지고 있는, 1000원짜리)을 의미하는 Lotto 클래스 구현
- isContained()메서드를 이용해 중복하는 번호가 있는지 확인

indiveloper committed:
refactor 변수명 변경

indiveloper committed:
refactor update README.md

기능 구현 목록 수정

indiveloper committed:
feat Validator 클래스 구현

InputView에서 입력 값의 유효성을 확인하는 메서드들을 Validator 클래스로 이동시킴.
Money 클래스를 테스트하는 중에 유효하지 않은 값을 테스트하려고 하는데 이미 유효성 확인을 마친 값을 입력받는 다는 전제하에 Money 클래스를 구현하다보니, Money 클래스 내에서는 유효성을 검사하는 기능이 없었음. 그래서 Validator 클래스에 checkInputMoney() 메서드를 만들어 InputView에서도 유효성 검사를 하고 Money에서도 유효성 검사를 할 수 있게 수정함.

indiveloper committed:
refactor LottoGenerator return 타입 수정

List<Number>를 반환하던 것에서 Lotto를 만들어 반환하는 것으로 수정

                                                                                                                                                                                                                               InputView에서 입력 값의 유효성을 확인하는 메서드들을 Validator 클래스로 이동시킴.
                                                                                                                                                                                                                               Money 클래스를 테스트하는 중에 유효하지 않은 값을 테스트하려고 하는데 이미 유효성 확인을 마친 값을 입력받는 다는 전제하에 Money 클래스를 구현하다보니, Money 클래스 내에서는 유효성을 검사하는 기능이 없었음. 그래서 Validator 클래스에 checkInputMoney() 메서드를 만들어 InputView에서도 유효성 검사를 하고 Money에서도 유효성 검사를 할 수 있게 수정함.

indiveloper committed:
refactor 변수명 변경

생성자의 매개변수 List<Number>의 변수명을 lotto에서 lottoNumbers로  변경

indiveloper committed:
refactor 유효성 검사를 Validator클래스를 이용해 하는 것으로 수정

Money 클래스에서 입력 값에 대해 유효성 검사를 하던 것을 Validator 클래스로 옮겨 Validator 클래스에서 함수를 호출해 유효성 검사 하는 것으로 수정

indiveloper committed:
feat PurchasedLottos 클래스 구현

사용자가 구입한 모든 로또의 리스트 List<Lotto>를 가지고 있는 클래스

indiveloper committed:
feat WinningLotto 클래스 구현

지난주 당첨 번호와 보너스 볼 넘버를 가지고 있는 WinningLotto 클래스 구현

indiveloper committed:
refactor 패키지 수정

model 패키지 생성

indiveloper committed:
feat Main 클래스 구현

로또 게임을 진행할 Main 클래스 구현

indiveloper committed:
feat 로또 구입 결과 출력

콘솔 출력을 담당하는 OutputView 클래스에 로또 구입 내역을 출력하는 printPurchasedLottoResult() 함수 구현

indiveloper committed:
refactor toString 오버라이드

Lotto의 정보를 알 수 있게끔 toString 오버라이드

indiveloper committed:
refactor 하드코딩한 값 상수화

하드코딩한 값을 상수화 함

indiveloper committed:
refactor printInputLottoNumberMessage 분리

수동 구매 로또 입력 메시지와 당첨 로또 입력 메시지를 한 번에 처리하던 printInputLottoNumberMessage()를 분리하여 수동 로또를 구매할 때와 지난주 당첨 번호를 입력할 때 각각 다른 메서드를 호출해서 실행하게 수정

indiveloper committed:
feat 보너스 번호 중복 체크 기능 추가

보너스 번호가 이미 입력 받은 지난주 당첨번호와 중복되는지 확인하는 기능 추가

indiveloper committed:
feat 구입한 모든 로또를 의미하는 PurchasedLottos 클래스 구현

수동으로 구입한 로또와 자동으로 구입한 로또 모두를 가지고 있는 PurchasedLottos클래스 구현

indiveloper committed:
feat Rank Enum 구현

로또 당첨 정보를 가지고 있는 Rank Enum 구현

indiveloper committed:
feat getter 추가

money getter 추가

indiveloper committed:
feat String을 Int로 바꿔주는 Convertor 추가

입력 받은 String값을 int로 바꿔주는 Convertor 추가

indiveloper committed:
refactor Scanner.nextInt()를 Scanner.nextLine()으로 변경

Convertor를 이용해 Scanner.nextLine()으로 받은 값을 int 타입으로 만들어주는 방식으로 코드 수정

indiveloper committed:
feat 로또 결과 출력

로또 결과 출력 기능 추가.

indiveloper committed:
feat Main 클래스 추가 구현

Lotto를 진행하는 Main클래스 추가 구현

indiveloper committed:
style 사용하지 않는 import 제거 및 줄맞춤

indiveloper committed:
feat Lotto 번호 정렬을 위한 Comparable 구현

indiveloper committed:
refactor String을 StringBuilder로 변경

문자열의 연산이 이루어지는 부분을 String에서 StringBuilder로 변경

phs1116 commented on src/main/java/lotto/domain/Main.java:
```diff
@@ -0,0 +1,85 @@
+import java.util.Map;
+
+public class Main {
+    public static void main(String[] args) {
```
Main이 너무 많은 책임들을 가지고 있는 것 같습니다.
대부분의 메서드에 있는 기능들은 각자 책임에 맞는 객체에 넘겨도 무방해보여요.:)
Main에서는 각 객체를 조합하도록만 하고 책임을 분리해보면 어떨까요? (judgeRank, calculatePrize 등)
그리고 외부의 Input과 도메인 로직이 한 메서드 내에 섞여있는 경향이 있는데 이를 명확하게 분리하도록 해보는것도 좋을 것 같아요. :)

soojinroh replied:
Main이 가지고 있던 메서드들을 각각의 객체로 분리시키고
도메인 로직에 포함되어 있던 input을 걷어내었습니다!

phs1116 commented on src/main/java/lotto/domain/model/Lotto.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    public List<Number> getLotto() {
+        return lotto;
```
Collections.unmodifiableList 참고해주세요!
이게 왜 필요한건지도 고민해보시면 좋을 것 같아요.:)

soojinroh replied:
이전에 Collections.unmodifiableList를 사용해보라는 피드백을 받아서 [내용 정리](https://github.com/soojinroh/Today_I_Learned/blob/master/TIL(2019-05-22)%20Collections.unmodifiablelist().md)까지 해봤는데 정작 코드 작성할 때는 사용하지 않고 있었던 것 같습니다..

Collections.unmodifiableList()를 사용해서 return하는 것으로 수정했습니다!

phs1116 commented on src/main/java/lotto/domain/model/Number.java:
```diff
@@ -0,0 +1,44 @@
+    private static final int MAX_LOTTO_NUMBER = 45;
+    private int lottoNumber;
+
+    public Number(int lottoNumber) {
```
이미 NumberSet을 통해 1차적으로 객체들을 캐시해놓고 있는데요,
Number에 대한 생성자를 공개하면 크게 의미 없어보입니다. :)
Number의 생성자를 은닉하고 static factory method를 통해 NumberSet에서 캐시한 값을 사용해보도록 해보면 어떨까요?

soojinroh replied:
피드백 주신 것 참고해서 수정해봤습니다!

phs1116 commented on src/main/java/lotto/domain/model/NumberSet.java:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    private static void checkLottoNumberRange(int lottoNumber) {
+        if (lottoNumber > MAX_LOTTO_NUMBER || lottoNumber < MIN_LOTTO_NUMBER) {
```
Number와 동일한 validation으로 보여요.
MAX_LOTTO_NUMBER, MIN_LOTTO_NUMBER, LOTTO_NUMBER_RANGE_EXCEPTION_MESSAGE 등 똑같이 의존해야하는 상수들도 많아보이고요.
Number 내에서 이러한 '캐시된 값'들을 관리해도 무방하지 않을까요? :)

soojinroh replied:
NumberSet과 Number에서 동일하게 사용하고 있는 변수와 메서드를 Number에서 캐싱하여 사용하는 것으로 수정하였습니다!

phs1116 commented on src/main/java/lotto/domain/model/Rank.java:
```diff
@@ -0,0 +1,59 @@
+            return SECOND;
+        }
+
+        for (Rank rank : values()) {
```
stream.filter.findAny.orElseThrow를 사용해보면 어떨까요? :)

soojinroh replied:
피드백 반영해서 수정해봤습니다!

phs1116 commented on src/main/java/lotto/domain/model/WinningLotto.java:
```diff
@@ -0,0 +1,21 @@
+
+public class WinningLotto {
+
+    private Lotto winningLotto;
```
컴포지션 👍

phs1116 commented on src/main/java/lotto/domain/Main.java:
```diff
@@ -0,0 +1,85 @@
+        List<Rank> ranks = new ArrayList<>();
+
+        for (Lotto lotto : purchasedLottos.getLottos()) {
+            ranks.add(Rank.valueOf(winningLotto.matchCount(lotto), winningLotto.matchBonusNumber(lotto)));
```
등수를 돌려주는 책임을 WinningLotto가 가질수 있지 않을까요?
Rank rank = winningLotto.match(lotto) 처럼요 :)

soojinroh replied:
WinningLotto에서 등수를 확인하고 상금을 결정하는 기능을 맡게 하였습니다!

phs1116 commented on src/main/java/lotto/domain/utils/Validator.java:
```diff
@@ -0,0 +1,63 @@
+import java.util.List;
+import java.util.Set;
+
+public class Validator {
```
Validator가 너무 전역적으로 체크 로직을 가지고 있어보여요.
모든 객체가 validation에 대한 결합도는 높지만 응집도는 떨어져보여요.
각 validation에 대한 책임은 각자 객체가 가지도록 하는게 좋을 것 같습니다.

soojinroh replied:
Validator의 기능을 각 객체에서 수행하는 것으로 수정하였습니다!

phs1116 commented on src/main/java/lotto/domain/utils/AutoLottoGenerator.java:
```diff
@@ -0,0 +1,25 @@
+            autoGeneratedLottoNumbers.add(NumberSet.of(key));
+        }
+
+        Collections.shuffle(autoGeneratedLottoNumbers);
```
Collections.shuffle에 직접적으로 의존하고 있는데요, 이 경우 자동 로또 생성에 대해서 값을 잘 돌려주는지 테스트가 힘듭니다.
 Collections.shuffler같이 테스트하기 힘든 부분에 대해서 느슨하게 의존하도록 해보면 어떨까요?
인터페이스를 중간에 가지고 의존하는 방법이 있을 것 같습니다. :)

soojinroh replied:
자동으로 로또를 생성하는 부분을 shuffle에 의존하지 않고 테스트 가능한 코드를 짜보려고 계속 고민해봤는데 답이 잘 나오지 않았습니다..


        // 수동 로또 생성
        while (manualLotto.size() < manualLottoSize) {
            InputView.printRemainedManualLotto(manualLottoSize, manualLotto);
            manualLotto.add(new Lotto(InputView.inputLottoNumber()));
        }

        // 자동 로또 생성
        int autoLottoAvailableSize = money.availablePurchaseCount() - manualLottoSize;
        while (autoLotto.size() < autoLottoAvailableSize) {
            autoLotto.add(new Lotto(ShuffledNumberGenerator.getShuffledNumbers()));
        }
   

이런 저런 방식을 생각해보다가 결국에는 Main에서 위와 같은 방식으로 구현했는데 LottoGenerator를 따로 만들자고 하니 사실 new Lotto()와 같이 생성해주는 역할밖에 하지 않는 것 같아 의미가 없어보였고
아래와 같이 LottoGenerator 인터페이스를 만들고 LottoGenerator를 구현하는 ManualLottoGerator, AuttoLottoGenerator를 만들고 각각의 클래스를 이용해서 Lotto를 생성하는 LottoFactory를 만드는 구조를 생각해보았을 때도 shuffle과 완벽히 분리되지 않는다는 느낌이 있습니다.  
  
```

public interface LottoNumbersGenerator {
    public List<Number> generate();
}

public class AutoLottoNumbersGenerator implements LottoNumbersGenerator {
    private ShuffleNumbers shuffleNumbers;

    public AutoLottoNumbersGenerator() {
        this(new ShuffleRandomNumers());
    }

    public AutoLottoNumbersGenerator(final ShuffleNumbers shuffleNumbers) {
        this.shuffleNumbers = shuffleNumbers;
    }

    public List<Number> generate() {
       List<Number> numbers = NumberSet.createNumbers();
         numbers=  shuffleNumbers.shuffle(numbers);
        return numbers.subList(0,Lotto.MAX_LOTTO_SIZE);
    }
}

public class LottosGenerateFactory {

    private final List<String> manualLotto;
    private final int availablePurchaseCount;

    public LottosGenerateFactory(final List<String> manualLotto, final int availablePurchaseCount) {
        this.manualLotto = new ArrayList<>(manualLotto);
        this.availablePurchaseCount = availablePurchaseCount;
    }

    public List<Lotto> generate(){
        List<Lotto> lottos = new ArrayList<>();

        for (final String lotto : manualLotto) {
            Lotto manualLotto = new Lotto(new ManualLottoNumbersGenerator(lotto).generate());
            lottos.add(manualLotto);
        }

        LottoNumbersGenerator lottosGenerator = new AutoLottoNumbersGenerator();
        for (int i = 0; i < availablePurchaseCount-lottos.size(); i++) {
            lottos.add(new Lotto(lottosGenerator.generate()));
        }

        return lottos;
    }
}

```

phs1116 replied:
LottoNumbersGenerator를 별도로 분리하는게 옳은 방향으로 생각되네요 :)
LottoNumbersGenerator 를 통해서 의존하기 때문에 shuffle과 직접 의존을 가지고 있지 않구요,
여기서 문제는 AutoLottoNumbersGenerator에 대한 의존성을 주입받는게 아닌 
코드 내부에서 직접 생성하다보니 의존을 느슨하게 가져가기 어려웠다는 점으로 보여요!
내부에서 직접 생성하면 LottosGenerateFactory는 shuffle을 사용하는 LottoNumbersGenerator에 직접적으로 의존하게 되니까요!
객체지향 원칙 중 DIP(의존성 역전 원칙)과 OCP(개방 폐쇄 원칙) 대해서 한번 살펴보시고 수정해보시면 좋을 것 같습니다. :)

phs1116 commented on src/main/java/lotto/domain/utils/ManualLottoGenerator.java:
```diff
@@ -0,0 +1,13 @@
+
+public class ManualLottoGenerator {
+
+    public static Lotto makeLotto(List<Number> lottoNumbers) {
```
그냥 생성자를 이용해도 상관없지 않을까요?
별도 객체를 생성하기에는 너무 책임이 없는 것 같습니다. :)

soojinroh replied:
생성자를 직접 사용하는 것으로 수정하였습니다!

phs1116 reviewed (changes requested):
안녕하세요! 리뷰어 화투입니다. :)
전체적으로 너무 잘 구현해주셨어요. 👍 
다만 결합도 및 객체의 책임에 대해서 피드백 몇가지 드렸는데요.
참고 부탁드릴게요!
궁금하신 점 있으시면 언제든지 연락주세요. :)

indiveloper committed:
refactor Collections.unmodifiableList를 이용해 반환값 수정

Collections.unmofiableList를 이용해 List를 변경하지 못하게 수정

indiveloper committed:
refactor ManualLottoGenerator 클래스 삭제에 따른 코드 수정

생성자를 호출해 Lotto 객체를 생성하는 역할만 하는 ManualLottoGenerator 클래스를 삭제하고 생성자를 직접 사용하는 것으로 코드 수정

indiveloper committed:
refactor getLotto()가 unmodifiable한 List를 반환함에 따른 코드 수정

indiveloper committed:
refactor LottoGenerator를 삭제하고 ShuffledNumberGenerator를 추가

AutoLottoGenerator와 ManualLottoGenerator를 삭제하고 lotto는 생성자로 생성함. 대신 ShuffledNumberGenerator를 이용해 랜덤한 6자리 숫자인 로또 번호만을 생성함.

indiveloper committed:
refactor Validator에서 수행하는 유효성 검사를 검사가 행해지는 클래스에서 하도록 수정

여러 클래스의 유효성 검사 기능을 Validator 클래스에 모아두었는데 이를 각자 클래스에서 수행하는 것으로 수정

indiveloper committed:
refactor Number와 NumberSet이 동일하게 사용하던 변수와 기능을 Number에서 캐싱하여 사용하는 것으로 수정

indiveloper committed:
refactor LottoGenerator 생성

LottoGenerator 인터페이스를 생성하고 AutoLottoGenerator와 ManualLottoGenerator에서 각각 LottoGenerator를 구현하여 자동/수동으로 로또를 생성하는 로직을 구현함.

indiveloper committed:
refactor for문으로 구현한 로직을 stream으로 수정

for문으로 구현한 로직을 stream.filter().findAy().orElseThrow()로 수정함.
로직에 변경은 없음.

indiveloper committed:
refactor Main에서 가지고 있던 메서드들을 각각의 객체로 분산

indiveloper committed:
refactor WinningLotto에서 순위를 판정하는 match기능 구현

indiveloper committed:
refactor 네이밍 변경

availablePurchaseTicketCount()에서 availablePurchaseCount()로 메서드 명 변경

indiveloper committed:
feat Number 클래스에 toString() 추가

indiveloper committed:
refactor 변수명 변경

~~Lottos였던 변수명을 ~~Lotto로 변경

indiveloper committed:
refactor getShuffledNumbers() 메서드 수정

기존에 numberPoll.subList()로 리턴해 사용하던 것을 새로운 List에 numberPoll.subList()를 넣어 리턴하는 것으로 수정.

indiveloper committed:
refactor input함수들에 조건 추가

checkManualLottoSize에 0장 이하일 경우 예외를 발생시키도록 조건 추가
printInputManualLottoMessage()에 수동 구입 장수가 0장 이하일 경우 메시지가 출력되지 않도록 조건 축가

indiveloper committed:
refactor 변수명 변경

Lottos에서 Lotto로 변수명 변경

indiveloper committed:
refactor 변수명 변경

Lottos에서 Lotto로 변수명 변경

indiveloper committed:
refactor LottoGenerator 클래스들 삭제

indiveloper committed:
refactor Main클래스 로직 수정

LottoGenerator를 없애고 Main 안에서 Lotto 생성을 수행

indiveloper committed:
refactor domain로직에 있던 출력문을 InputView 클래스의 메서드로 추출

indiveloper committed:
refactor Lotto 생성 부분 수정

makeAutoLotto()와 makeManualLotto()에서 수행하던 Lotto생성 기능을 Main메서드에서 담당하는 것으로 수정

phs1116 commented on src/main/java/lotto/domain/model/NumberSet.java:
```diff
@@ -0,0 +1,24 @@
+import static lotto.domain.model.Number.MAX_LOTTO_NUMBER;
+import static lotto.domain.model.Number.checkLottoNumberRange;
+
+public class NumberSet {
```
NumberSet보다는 Number 객체를 생성하는 역할이니  Factory같은 이름이 더 적합할 것 같네요. :)

phs1116 commented on src/main/java/lotto/domain/Main.java:
```diff
@@ -0,0 +1,49 @@
+            autoLotto.add(new Lotto(ShuffledNumberGenerator.getShuffledNumbers()));
+        }
+
+        purchasedLotto.addLotto(manualLotto);
```
purchaseLotto에 대한 부분들도 기존대로 한번 더 추상화 할 수 있을것같네요 :)
로또 생성에 대한 부분이 다른 코드들과 추상화 레벨이 맞지 않는 것 같아요!

phs1116 reviewed (approved):
안녕하세요! 리뷰어 화투입니다.
피드백 반영하시느라 정말 수고 많으셨습니다. :)
이번 단계 머지할게요.
코멘트 중 고민사항으로 남겨주신 점과 조금 더 공부하고 개선할 수 있는 점 추가로 달아놨는데요,
웹쪽 개발 진행하시면서도 같이 고민 부탁드릴게요. :)
